extra/bunslog/bunslog.go

```diff
@@ -0,0 +1,113 @@
+package bunslog
+
+// bunslog provides logging functionalities for Bun using slog.
+// This package allows SQL queries issued by Bun to be displayed using slog.
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/uptrace/bun"
+)
+
+// Option is a function that configures a QueryHook.
+type Option func(*QueryHook)
+
+// WithQueryLogLevel sets the log level for general queries.
+func WithQueryLogLevel(level slog.Level) Option {
+	return func(h *QueryHook) {
+		h.queryLogLevel = level
+	}
+}
+
+// WithSlowQueryLogLevel sets the log level for slow queries.
+func WithSlowQueryLogLevel(level slog.Level) Option {
+	return func(h *QueryHook) {
+		h.slowQueryLogLevel = level
+	}
+}
+
+// WithErrorQueryLogLevel sets the log level for queries that result in an error.
+func WithErrorQueryLogLevel(level slog.Level) Option {
+	return func(h *QueryHook) {
+		h.errorLogLevel = level
+	}
+}
+
+// WithSlowQueryThreshold sets the duration threshold for identifying slow queries.
+func WithSlowQueryThreshold(threshold time.Duration) Option {
+	return func(h *QueryHook) {
+		h.slowQueryThreshold = threshold
+	}
+}
+
+// SlogFunc is a type that defines the signature for slog logging functions.
+type slogFunc func(ctx context.Context, msg string, attrs ...any)
+
+// QueryHook is a hook for Bun that enables logging with slog.
+// It implements bun.QueryHook interface.
+type QueryHook struct {
+	queryLogLevel      slog.Level
+	slowQueryLogLevel  slog.Level
+	errorLogLevel      slog.Level
+	slowQueryThreshold time.Duration
+	slogFns            map[slog.Level]slogFunc
+	now                func() time.Time
+}
+
+// NewQueryHook initializes a new QueryHook with the given options.
+func NewQueryHook(opts ...Option) *QueryHook {
+	slogFns := map[slog.Level]slogFunc{
+		slog.LevelDebug: slog.DebugContext,
+		slog.LevelInfo:  slog.InfoContext,
+		slog.LevelWarn:  slog.WarnContext,
+		slog.LevelError: slog.ErrorContext,
+	}
+
+	h := &QueryHook{
+		queryLogLevel:     slog.LevelDebug,
+		slowQueryLogLevel: slog.LevelWarn,
+		errorLogLevel:     slog.LevelError,
+		slogFns:           slogFns,
+		now:               time.Now,
+	}
+
+	for _, opt := range opts {
+		opt(h)
+	}
+
+	return h
+}
+
+// BeforeQuery is called before a query is executed.
+func (h *QueryHook) BeforeQuery(ctx context.Context, event *bun.QueryEvent) context.Context {
+	return ctx
+}
+
+// AfterQuery is called after a query is executed.
+// It logs the query based on its duration and whether it resulted in an error.
+func (h *QueryHook) AfterQuery(ctx context.Context, event *bun.QueryEvent) {
+	duration := h.now().Sub(event.StartTime)
+	if event.Err != nil && !errors.Is(event.Err, sql.ErrNoRows) {
+		msg := fmt.Sprintf("[bunslog] error query: %s", event.Query)
```

Comment:
Can we use structured logging here instead? I.e. `slog.String("query", event.Query)`

Reply:
@vmihailenco 
Yes, it's possible to use structured logging with `slog.String("query", event.Query)`. Would you like to omit the default `msg` field in that case? 

Regarding the expected log format, are you looking for something like this?
```json
{"time":"xxxx-xx-xxTxx:xx:xx+xx:xx","level":"DEBUG", "query":" SELECT `user`.`id`, `user`.`name`, `user`.`email` FROM `users`", "operation":"SELECT","duration":"3s"}
```
Or would you prefer this format?
```json
{"time":"xxxx-xx-xxTxx:xx:xx+xx:xx","level":"DEBUG", "msg":"[bunslog] query", "query":" SELECT `user`.`id`, `user`.`name`, `user`.`email` FROM `users`", "operation":"SELECT","duration":"3s"}
```

Reply:
>Would you like to omit the default msg field in that case?

I think so, yes.

>Adding an additional field, such as status is_slow, could help make this distinction clear.

For errors we should just add `slog.Any("error", event.Err)`. For slow queries I would not do anything special.

Reply:
I've omitted `msg` field and have fixed to allow for custom formats to be passed in as an option.

https://github.com/hirasawayuki/bun/blob/433ad6c04c293a570f3693fe055f2f45c7956964/extra/bunslog/bunslog.go#L93-L98